File: 2021/go/day6/part2/part2.go
package part2

import (
	"strconv"
	"strings"
)

var Days = 256

func Solve(input string) string {
	result := 0
	for _, cycle := range strings.Split(input, ",") {
		n, _ := strconv.Atoi(cycle)
		result += a[Days-n]
	}
	return strconv.Itoa(result)
}

var a [257]int

func init() {
	// dp: f(n) = f(n-7) + f(n-9)
	for i := 1; i <= 256; i++ {
		a[i] = f(i-7) + f(i-9)
	}
}

func f(n int) int {
	if n > 0 {
		return a[n]
	}
	return 1
}

File: 2021/elixir/apps/day6/lib/part2.ex
defmodule Day6Part2 do
  @moduledoc """
  Day6 - Part2
  """

  def solve(input) do
    input
    |> String.split(",")
    |> Enum.map(&descendants(String.to_integer(&1), 256))
    |> Enum.sum()
  end

  defp descendants(cycle, days) do
    # dp: f(n) = f(n-7) + f(n-9)
    state =
      Enum.reduce(1..days, %{}, fn day, state ->
        Map.put(state, day, Map.get(state, day - 7, 1) + Map.get(state, day - 9, 1))
      end)

    Map.get(state, days - cycle)
  end
end
